### PR TITLE
Issue #1555 WWW-Authenticate realm header ordering

### PR DESCRIPTION
**Pull request by lachlan-roberts**
Changed regex for AUTHENTICATE_PATTERN so that the realm doesn't need to appear as first parameter. #1555

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

### CODE REVIEW

**Comment by lachlan-roberts** [comment]
I don't think there is any easy way to change the ordering of the authentication parameters to test it, or should i just be writing tests for the regex itself?

**Comment by joakime** [comment]
We could rip out the Authentication header parsing into a unique class or method that we can use in the testing.
We would pass in the various exemplar headers and assert the results.

File: jetty-client/src/main/java/org/eclipse/jetty/client/AuthenticationProtocolHandler.java [hunk 1]
```diff
@@ -42,7 +42,7 @@
 {
     public static final int DEFAULT_MAX_CONTENT_LENGTH = 16*1024;
     public static final Logger LOG = Log.getLogger(AuthenticationProtocolHandler.class);
-    private static final Pattern AUTHENTICATE_PATTERN = Pattern.compile("([^\\s]+)\\s+realm=\"([^\"]*)\"(.*)", Pattern.CASE_INSENSITIVE);
+    private static final Pattern AUTHENTICATE_PATTERN = Pattern.compile("([^\\s]+)\\s(.*)realm=\"([^\"]*)\"(.*)", Pattern.CASE_INSENSITIVE);
```
Comment (sbordet) [inline jetty-client/src/main/java/org/eclipse/jetty/client/AuthenticationProtocolHandler.java]: This regexp also matches:
```
Digest qop="auth", blahrealm="thermostat", nonce="1523430383"
```

File: jetty-client/src/main/java/org/eclipse/jetty/client/AuthenticationProtocolHandler.java [hunk 2]
```diff
@@ -241,8 +241,8 @@ private void forwardFailureComplete(HttpRequest request, Throwable requestFailur
                 if (matcher.matches())
                 {
                     String type = matcher.group(1);
-                    String realm = matcher.group(2);
-                    String params = matcher.group(3);
+                    String realm = matcher.group(3);
+                    String params = matcher.group(2) + matcher.group(4);
```
Comment (sbordet) [inline jetty-client/src/main/java/org/eclipse/jetty/client/AuthenticationProtocolHandler.java]: With the example above, `params` becomes:
```
qop="auth", blah, nonce="1523430383"
```

In the good case, it becomes:
```
qop="auth", , nonce="1523430383"
```

I'm not sure the rest of the code is able to parse the double comma properly.

Comment (lachlan-roberts) [reply jetty-client/src/main/java/org/eclipse/jetty/client/AuthenticationProtocolHandler.java]: i checked that and the rest of the code seems to be fine handling the double comma
in fact if you add this line in after setting params it still passes all the tests
`params = ",  , , , ,  , ," + params + ",  , , , ,, , ,  ,";`

File: jetty-client/src/main/java/org/eclipse/jetty/client/AuthenticationProtocolHandler.java [hunk 3]
```diff
@@ -42,7 +42,7 @@
 {
     public static final int DEFAULT_MAX_CONTENT_LENGTH = 16*1024;
     public static final Logger LOG = Log.getLogger(AuthenticationProtocolHandler.class);
-    private static final Pattern AUTHENTICATE_PATTERN = Pattern.compile("([^\\s]+)\\s+realm=\"([^\"]*)\"(.*)", Pattern.CASE_INSENSITIVE);
+    private static final Pattern AUTHENTICATE_PATTERN = Pattern.compile("([^\\s]+)\\s(.*,\\s*)?realm=\"([^\"]*)\"\\s*,?\\s*(.*)", Pattern.CASE_INSENSITIVE);
```
Comment (sbordet) [inline jetty-client/src/main/java/org/eclipse/jetty/client/AuthenticationProtocolHandler.java]: I tested this regexp against `Digest  realm="foo"` and it does not pass (note 2 spaces between `Digest` and `realm`.
It passes if you modify the regexp to be `([^\\s]+)\\s+...` (note the `+` after the non grouped `\\s`).

File: jetty-client/src/main/java/org/eclipse/jetty/client/AuthenticationProtocolHandler.java [hunk 4]
```diff
@@ -241,8 +241,13 @@ private void forwardFailureComplete(HttpRequest request, Throwable requestFailur
                 if (matcher.matches())
                 {
                     String type = matcher.group(1);
-                    String realm = matcher.group(2);
-                    String params = matcher.group(3);
+                    String realm = matcher.group(3);
+                    String params;
+                    if(matcher.group(2) != null)
```
Comment (sbordet) [inline jetty-client/src/main/java/org/eclipse/jetty/client/AuthenticationProtocolHandler.java]: Put a space between `if` and `(`.
Introduce a local variable for `matcher.group(2)`, maybe named `beforeRealm`.

**Pull request merged**